posts/2017-02-13-more-types-than-classes.md

```diff
@@ -0,0 +1,702 @@
+---
+layout: post
+title: There are more types than classes
+
+meta:
+  nav: blog
+  author: S11001001
+  pygments: true
+---
+
+As programmers, we are very incautious with our use of the word
+“type”. The concept of “type” is sufficiently abstract and specific
+that we are tempted to understand it by analogy, so much that we begin
+to confuse analogy with sameness.
+
+The colloquial “runtime type”, a fair approximation of “class”, makes
+it tempting to equate _types_ with “classes, interfaces, traits, that
+sort of thing”, which I will name _classes_ for the rest of this
+article. But they aren’t the same. The type system is much richer and
+more interesting than the class system, even in Java.
+
+To appreciate this richness, we must stop thinking of types as classes
+and stop drawing conclusions from that weak analogy. Luckily, the
+compiler will readily reveal how unlike classes types are, if we ask
+it some simple questions.
+
+## One value with class, many variables with type
+
+```scala
+val greeting: String = "hi there!"
+```
+
+Here I have constructed a `String` and assigned it to a variable. (I
+have also constructed the `char` array in the `String` and various
+other details, but immediately handed those off to the `String` and
+forgotten about them.) This value has class `String`. It has several
+classes, really.
+
+1. `String`
+2. `java.io.Serializable`
+3. `CharSequence`
+4. `Comparable[String]`
+5. `Object`/`AnyRef`
+
+That seems like a lot of classes for one value. And they are genuine
+classes of `greeting`, though 2-5 are all implied by #1.
+
+`greeting` also has all five of these _types_. We can ask the compiler
+to verify that this _type_ truth holds, entirely separately from the
+class truth.
+
+```scala
+scala> (greeting: String, greeting: java.io.Serializable,
+        greeting: CharSequence, greeting: Comparable[String],
+        greeting: AnyRef)
+res3: (String, java.io.Serializable, CharSequence,
+       Comparable[String], AnyRef) =
+  (hi there!,hi there!,hi there!,hi there!,hi there!)
+```
+
+So we have exhausted the classes, but aren’t quite done with types.
```

Comment:
Could you also include (either here, or later) a mention of the singleton literal type, as distinct from the value type? Whilst I don't think it adds all that much, "we...aren't quite done with types" suggests that we had only missed one type. It might then be necessary to note that the singleton literal types arising from two different (but equal) values still conform, even though their value types don't... (You may equally decide, quite justifiably, that the subtleties of Scala's inference and type-widening are too complicated for this article, and skip them!)